### ISSUE TITLE
Should the capnp compiler work when compiled without RTTI?

### ISSUE THREAD

**Issue by nlyan** [capnproto/capnproto#1227]
First of all, thanks so much for the continued work on Cap'n'Proto. I evaluated it years ago but it's really come a long way since, and am now planning to use it in a project.

Is it expected that the capnp compiler work when compiled without RTTI? 

I'm currently building the library and compiler on MSVC 2017 with /GR-, and using the CMake integration on a project, but I'm getting "std::terminate() called with no exception" when CMake tries to a compile my schema file.

I've confirmed that when /GR- is removed from the capnproto library build everything works just fine.

If the compiler requires RTTI but the library doesn't, then I will have to stop using the CMake integration, as I'll need it to use the compiler built with RTTI to compile the schema files, and then link against the RTTI-less library. I've tested the library without RTTI (on Linux) and it seemed to work fine.

**Comment by kentonv** [comment]
Our release tests do include building the whole project -- including compiler -- with `-fno-rtti -fno-exceptions` on GCC and Clang, and verifying that that all works. However, we don't test MSVC in this configuration, and we don't test the two flags separately (we assume most people who want to disable RTTI also want to disable exceptions). While I would _expect_ that disabling one and not the other should work fine, and I would expect the behavior on MSVC to match GCC/Clang in this regard, since it isn't tested it could very easily be broken.

The compiler's reporting of errors may be worse when exceptions are disabled. (But I wouldn't expect RTTI to matter.)

Note that no-rtti/no-exceptions modes are often broken between releases. If you need to build without RTTI, then you definitely need to stick to a release version.

Note also that the cmake integration is primarily maintained by contributors that aren't core maintainers of Cap'n Proto. I personally don't know cmake. I would think it should allow specifying an external capnp binary to use, and if not, I'd accept a PR adding that support.

**Comment by nlyan** [comment]
Hi Kenton,

Thanks for the quick reply.

I did a debug build and stuck the debugger on it, and found that this is broken due to the assumptions laid out here:

https://github.com/capnproto/capnproto/blob/master/c%2B%2B/src/kj/memory.h#L510

and probably here too:

https://github.com/capnproto/capnproto/blob/master/c%2B%2B/src/kj/memory.h#L89

The Itanium C++ ABI used on open platforms does indeed contain the necessary offset in the vtable to allow for dynamic_cast to void* without RTTI, but this is not possible on MSVC.

I don't know if this effects just the compiler, or the library as a whole, but I think this means MSVC builds without RTTI (regardless of exception use) are broken.

**Comment by kentonv** [comment]
I'm afraid that `dynamic_cast<void*>` is very much needed in the implementation of `kj::Own`, which is used all over in KJ and Cap'n Proto. It matters for any multiple-inheritance scenario, not just diamond dependency graphs.

The problem is this: Say you have `class Sub: public Super1, public Super2`. You allocate a `kj::Own<Sub>`. Later, you cast the pointer to `kj::Own<Super2>`. Later still, you drop this pointer, so the object needs to be destroyed.

On destruction, we need to figure out the pointer to the start of the object -- i.e. to the instance of `Sub` -- so that we can pass it to the disposer. The pointer we have is to `Super2`. Unfortunately, these pointers are numerically different; `Super2` is offset within `Sub` by the size of `Super1`. `dynamic_cast<void*>` gives us a pointer to the start of the object, no matter which superclass we start from.

The information needed is definitely present in the vtable, because `operator delete` has to do exactly the same thing when it is invoked, in order to decide what pointer to pass to `free()`. However, we don't use `operator delete`, because `kj::Own` offers as a feature the ability to specify a custom disposer, with dynamic dispatch.

In order to implement this without `dynamic_cast<void*>`, we would instead have to increase the size of every `kj::Own<T>` to include a base pointer for the object, so that we don't have to derive it. This would have the nice benefit that we could allow upcasting on non-polymorphic types, which is currently forbidden. But increasing the size of all `kj::Own`s would be unfortunate.

Since this only affects MSVC in no-RTTI mode, I suppose someone could implement some #ifdef hackery that implements this approach specifically for that target, while keeping the code as-is everywhere else.

But I have to ask... is compiling with RTTI really that bad?

**Comment by nlyan** [comment]
> The information needed is definitely present in the vtable, because operator delete has to do exactly the same thing when it is invoked

This is incorrect even on open platforms. If Sub, Super1 and Super2 are all polymorphic then the way delete is implemented is that the compiler will generate a special vtable for a "Super2-inside-a-Sub" that has the exact same layout as the vtable for an ordinary Super2. This vtable will contain virtual function pointers to a bunch of thunks that perform the this pointer adjustment before calling the implementation of the Sub destructor:

Take a look at: https://gcc.godbolt.org/z/PnGqP4aP6

Calling disposeOfSuper2() on a Sub* in this example is of course completely safe.

You can see at the bottom, on line 40 of the asm, under "vtable for Sub" that the "non-virtual thunk to Sub::~Sub() [deleting destructor]" entry is added to the vtable. 

This is defined on line 22, where the offset from Super1 to Super2 is effectively compiled statically as "sub     rdi, 56".

```
non-virtual thunk to Sub::~Sub() [deleting destructor]:
        sub     rdi, 56
        jmp     .LTHUNK1
```

56 bytes are subtracted here: 48 bytes of 'somethingBig' in Super1 plus an additional vtable pointer. 

There's therefore no requirement for this offset to exist in the vtable itself just to implement destructors, and even though it does exist under the Itanium ABI  to support dynamic_cast<void*> efficiently, it's not used for this purpose.

What I think should happen here is that when you cast from Own<Sub> to Own<Super2> the disposer should be wrapped to properly static_cast the void* back to Super2* before then calling the *original* disposer. This unfortunately means a heap allocation to support this cast, unless some constexpr/type-traits magic can optimize it away in the single-inheritance case.

**Comment by nlyan** [comment]
It looks like Mozilla ran in to this same issue at some point:

https://bugzilla.mozilla.org/show_bug.cgi?id=976372

**Comment by nlyan** [comment]
I see three alternative solutions here:

1. static_assert for MSVC builds where RTTI is disabled, so the library can't be built in this configuration.
2. When casting from Own\<Derived\> to Own\<Base\> allocate and construct a new disposer that static_cast's the void* (Base*)  back to Derived* before then calling the *original* disposer. This unfortunately means a heap allocation to support this cast, unless some constexpr/type traits magic can be used to  optimize it away on the happy path (single inheritance). This slightly defeats the point of using Own<T> in the first place, although it only applies to conversions and not ordinary moves.
3. Make Own<Base> fatter by having it contain both a Base* and a void*. When an Own\<Derived\> is converted to an Own\<Base\> the void* is always preserved so that it can be passed to the disposer. This is exactly how std::shared_ptr works and is why it is 2 pointers in size. Unfortunately because Own<T*> carries the Disposer* as well, this will make Own<T> 3 pointers in size.

If option 3 depends on the compiler configuration, it has potential ABI concerns if some code is compiled without RTTI and other code is compiled with RTTI.

**Comment by englercj** [comment]
I've hit this issue as well.

> But I have to ask... is compiling with RTTI really that bad?

If it is just the tool, but not the libs then it is probably fine. But having RTTI enabled in the runtime of our application (games) is just a non-starter.

@nlyan did you work around this issue or end up creating a patch? I'd really like to use capnp, but if RTTI is required for the runtime I'm afraid I'll have to find another solution.

**Comment by kentonv** [comment]
> But having RTTI enabled in the runtime of our application (games) is just a non-starter.

Why is that? I'd like to understand. AFAIK RTTI doesn't affect performance unless you use it. It may add a little bit to the code footprint (for extra vtable data) -- is that the problem?

**Comment by englercj** [comment]
I don't think RTTI has any performance implications when unused, but the extra vtable data is an issue. The extra data can be helpful to attackers looking to make cheats. Although disabling it obviously doesn't stop anyone, enabling it can make it much easier to develop cheats. All game clients I've worked on disabled RTTI and exceptions, including my current project.

**Comment by kentonv** [comment]
Wow OK, that's interesting! I never would have guessed that. I suppose if you had some tooling that could remove the class names, it wouldn't be an issue, but I can see how it's much easier to just turn it off entirely.
